Missing bits from maistra-2.1

tedjpoole opened the pull request.

ManirajDeivendran committed:
Add luajit2 support in maistra/envoy (#111)

Conflicts:
	bazel/foreign_cc/BUILD
	bazel/repositories.bzl
	bazel/repository_locations.bzl

luksa committed:
MAISTRA-2687 Fix spiffe certificate verification (#116)

* MAISTRA-2687 Fix spiffe certificate verification

Previously, when copying the X509 store context, the code erroneously copied the output certificate chain instead of the input (untrusted) chain to the new context. As the output certificate chain is set only after you run X509_verify_cert(), this chain was always empty. Thus, the code only worked properly when the leaf certificate was signed by one of the certificates in the trust bundle (e.g. when the leaf certificate was signed directly by the root certificate). In cases involving intermediate certificates, the leaf certificate was only trusted if all the intermediate certificates were in the trust bundle. Intermediate certificates sent by the client were ignored.

When copying the X509 store context, the input intermediate certificates should be retrieved by calling X509_STORE_CTX_get0_untrusted(), not X509_STORE_CTX_get0_chain().

* Add test

* Fix test

* Release references to certs

Conflicts:
	source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

Signed-off-by: Marko Lukša <marko.luksa@gmail.com>

twghu committed:
Adding .gitleaks for identifying test data secrets

Signed-off-by: Tim Walsh <temporal.differential@gmail.com>

twghu committed:
Added comment and removed duplicates

Signed-off-by: Tim Walsh <temporal.differential@gmail.com>

tedjpoole committed:
Removed OCSP response expiry check

tedjpoole committed:
Fixed UMR in ALPN protocol name extraction

oschaaf commented on source/extensions/filters/listener/tls_inspector/tls_inspector.cc:
```diff
@@ -228,14 +228,20 @@ std::vector<absl::string_view> Filter::getAlpnProtocols(const unsigned char* dat
+    uint8_t protocol_length = data[i++];
+
+    if (protocol_length == 0) {
+      ENVOY_LOG(debug, "tls inspector: ALPN protocol length is zero");
```
I wonder if emitting a warning in combination with a switch to `ENVOY_LOG_PERIODIC_TO_LOGGER` [1] would be better here. wdyt? 

[1] https://github.com/envoyproxy/envoy/blob/e727e1f50b33730a3a4fc061f95bcdd6d0c48768/source/common/common/logger.h#L541

tedjpoole replied:
I wasn't too sure about logging level to use. I had a look around and it seemed to me like warning/error log messages are generally used to report on the (mal)functioning of envoy itself, rather than to report on erroneous content flowing through envoy, but I'll change it as suggested.

oschaaf replied:
The thing is that with regular logging this could result in far too much log spew. But i think that hitting these edge cases is worth surfacing to the user. With the periodic log function, we don't be flooding the logs while still surfacing that this stuff is happening. That was my reasoning.

tedjpoole replied:
Done

oschaaf commented on source/extensions/filters/listener/tls_inspector/tls_inspector.cc:
```diff
@@ -228,14 +228,20 @@ std::vector<absl::string_view> Filter::getAlpnProtocols(const unsigned char* dat
-    ++i;
-    absl::string_view protocol(str.substr(i, protocol_length));
-    protocols.push_back(protocol);
+    uint8_t protocol_length = data[i++];
```
So the format is that data is prefixed with a single byte, which represents the number of data bytes, correct? Can you add code-level comment to the method to document Filter::getAlpnProtocols while we are at it?

Tangentially related, but imho out of scope for now -- I'd love it if we could refactor this method into it's own unit of compilation under `/maistra/source/tls/alpn_util.cc` or some such.. and then we could have unit tests for this in `/maistra/source/tls/test_alpn_util.cc`.  That would prove correctness, tease out issues when run under asan, as well as serve as semi-documentation on assumptions and edge cases etc.

tedjpoole replied:
Added comment to explain the format of the data being unmarshalled.

oschaaf replied:
👍 thank you, resolving.

oschaaf commented on source/extensions/filters/listener/tls_inspector/tls_inspector.cc:
```diff
@@ -228,14 +228,20 @@ std::vector<absl::string_view> Filter::getAlpnProtocols(const unsigned char* dat
   }
 
-  absl::string_view str(reinterpret_cast<const char*>(data));
   for (unsigned int i = 0; i < len;) {
```
should we put a cap on the value of len that we are willing to accept?

tedjpoole replied:
At the point where getAlpnProtocols() is called the buffer has already been allocated and populated off the wire, so why would we reject what's in it?

oschaaf replied:
makes sense resolving

oschaaf commented on source/extensions/filters/listener/tls_inspector/tls_inspector.cc:
```diff
@@ -228,14 +228,20 @@ std::vector<absl::string_view> Filter::getAlpnProtocols(const unsigned char* dat
+    }
+
+    if (protocol_length > (len - i)) {
+      ENVOY_LOG(debug, "tls inspector: ALPN protocol length is too long");
```
ditto

tedjpoole replied:
done

oschaaf commented on source/extensions/filters/listener/tls_inspector/tls_inspector.cc:
```diff
@@ -228,14 +228,20 @@ std::vector<absl::string_view> Filter::getAlpnProtocols(const unsigned char* dat
+    }
+
+    protocols.emplace_back(reinterpret_cast<const char*>(data + i), protocol_length);
     i += protocol_length;
```
readability nit: afaict this increment could move into the usual position in the for() loop on the premise that we declare it outside of that and initialize it to zero -- but that probably is just personal preference

oschaaf commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -737,10 +737,6 @@ ServerContextImpl::ServerContextImpl(Stats::Scope& scope,
       if (!response->matchesCertificate(*tls_context_.cert_contexts_[i].cert_chain_)) {
         throw EnvoyException("OCSP response does not match its TLS certificate");
       }
-      if (response->isExpired()) {
```
Did we verify upstream also stripped this? Or is this specific to us?

tedjpoole replied:
This is specific to us and the decision to remove it was made in discussion with @twghu

oschaaf replied:
great, thanks for sharing context. perhaps leave note for our future selves in the code? e.g. something like:

```
// MAISTRA - removed "if (response->isExpired()" handling, because .. 
```

tedjpoole replied:
Reinstated this check because it looks like it has some value. What do you think @twghu? Should I also add it upstream?

twghu replied:
The expiry check if() was added 
on 13/4/21 https://github.com/maistra/envoy/commit/412c091b302cc688bf3269022bf679bced828b2f#diff-2b39f212c2f760ae547351a46f92ad08c0ec84c556d678ebe9388b9d008e7313

and was removed on 20/5/21
https://github.com/maistra/envoy/commit/963afc8094dc2c17bd24526f2de11841a5904b1b

The last commit should indicate reason for removal, a sanity check would be a good idea.

oschaaf replied:
So digging into this, it seems the check is (also) done in [ServerContextImpl::ocspStapleAction](https://github.com/maistra/envoy/blob/0428980c9210b2c7de98eb64e73fcfc526c96426/source/extensions/transport_sockets/tls/context_impl.cc#L948).  So I think we can remove it.

oschaaf commented on source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:
```diff
@@ -165,9 +165,9 @@ int SPIFFEValidator::doVerifyCertChain(X509_STORE_CTX* store_ctx,
-  // Note that there's no api to copy crls from one store_ctx to another; the assumption is that
-  // neither X509_V_FLAG_CRL_CHECK nor X509_V_FLAG_CRL_CHECK_ALL verify_params are not used.
-  // Should this change, consider opening up X509_STORE_CTX struct, which is internal atm.
+  // Note that there's no api to copy crls from one store_ctx to another; the assumption is that 
```
Are we sure we included the rest of Marko his fix for this validation?

tedjpoole replied:
Yes, it was only the change to the comment that didn't make it to 2.2

oschaaf reviewed (changes requested):
Nice work, like how readable `getAlpnProtocols` got. Left a few nits/questions/comments.

jwendell commented on bazel/foreign_cc/luajit2.patch:
```diff
@@ -0,0 +1,98 @@
+ # unwinding are not affected -- the assembler part has frame unwind
+ # information and GCC emits it where needed (x64) or with -g (see CCDEBUG).
+-CCOPT= -O2 -fomit-frame-pointer
++CCOPT= -O2 -fomit-frame-pointer -fPIC -fPIE -fplugin=annobin -fstack-protector-strong -fstack-protector-all
```
We removed this in a recent pr

jwendell replied:
This change needs to be removed otherwise it will fail and we will only notice it on ppc builds, not on prow (x86). This is an issue: We currently don't have a way to try builds on non-x86 platforms before the pr is merged.

tedjpoole replied:
Done. It builds & tests ok (on x86) using the `--//source/extensions/filters/common/lua:luajit2=1` option.

jwendell replied:
Awesome, thanks.

jwendell commented on bazel/repositories.bzl:
```diff
@@ -1006,6 +1006,9 @@ def _com_github_luajit2_luajit2():
     external_http_archive(
         name = "com_github_luajit2_luajit2",
         build_file_content = BUILD_ALL_CONTENT,
+        patches = ["@envoy//bazel/foreign_cc:luajit2.patch"],
```
Are you sure? I think this patch is already present on 2.2

knm3000 replied:
It is currently not present in 2.2: some luajit2 changes have not been merged from 2.1 into 2.2. This PR contains all missing luajit2 code that is needed to build envoy/proxy with luajit2.

jwendell replied:
True. I just hit this in a ppc product build. The build fails because of this.

tedjpoole replied:
@jwendell just to be clear, do you mean that you want this left in?

jwendell replied:
Yes. This is missing in 2.2 indeed.

You just need to get rid of the change to the flags mentioned above
```diff
-CCOPT= -O2 -fomit-frame-pointer
+CCOPT= -O2 -fomit-frame-pointer -fPIC -fPIE -fplugin=annobin -fstack-protector-strong -fstack-protector-all
```

oschaaf commented on source/extensions/filters/listener/tls_inspector/tls_inspector.cc:
```diff
@@ -228,15 +228,25 @@ std::vector<absl::string_view> Filter::getAlpnProtocols(const unsigned char* dat
+  // string is preceded by 1 byte that gives it's length (excluding the length
+  // byte itself). No string can be zero length, and the last string must not
+  // overrun or underun the buffer.
+  uint8_t protocol_length;
```
shall we init protocol_length to 0? :-)

tedjpoole replied:
It didn't look necessary to me because it does get initialised before use.

oschaaf replied:
Sorry if I'm a little dense, but where does protocol_length get initialised? 
(From what I can tell `i += protocol_length` is the first statement that mentions it?)

tedjpoole replied:
The `i += protocol_length` statement doesn't get executed until after the first run through the loop, so the `protocol_length = data[i++]` statement is the first to be executed.

jwendell replied:
evil :D

oschaaf replied:
Aaah yes, I missed that; thanks for explaining. Resolving this..

oschaaf replied:
though thinking about it, perhaps a one line comment about this won't hurt here just to make it obvious to future readers?

jwendell replied:
or just initialize it... compiler might optimize it anyway, so...

We are also doing other evil stuff like changing the for loop variable in the middle of the loop. It might work, but my personal opinion is that in such cases we could just use while instead of for.

oschaaf replied:
So I mulled that over a bit, and now I agree on that I'd prefer a while loop as well. It seems like we're going through hoops to try and fit what we are doing into the original for loop construction, which I think started us off on the wrong foot: a while loop seems to be a more natural intuitive candidate here. @tedjpoole sorry about the flipflopping :-) curious to hear what you think?

oschaaf replied:
thanks for following up, I like the while loop better. Resolving
/cc @jwendell

oschaaf commented on source/extensions/filters/listener/tls_inspector/tls_inspector.cc:
```diff
@@ -228,15 +228,25 @@ std::vector<absl::string_view> Filter::getAlpnProtocols(const unsigned char* dat
+  // The data buffer contains a sequence of non-terminated strings, where each
+  // string is preceded by 1 byte that gives it's length (excluding the length
+  // byte itself). No string can be zero length, and the last string must not
+  // overrun or underun the buffer.
```
perfect, thank you for adding this. 
just one nit: s/underun/underrun/

tedjpoole replied:
Done

tedjpoole committed:
Reinstated OCSP response expiry check

This reverts commit 47b70577d8f7dc6ce7b817b7c098b9d9919e65f0.

tedjpoole committed:
Fixed UMR in ALPN protocol name extraction

Converted the for loop into a while loop for readability

oschaaf reviewed (approved):
modulo pending comments from others, this LGTM

tedjpoole committed:
Fixed luajit2 patch

Tested t builds (on x86) with --//source/extensions/filters/common/lua:luajit2=1 switch
